processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/CsvInput.java

```diff
@@ -343,41 +349,57 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
     } catch (NumberFormatException exc) {
       numberOfNodes = NUM_CORES_DEFAULT_VAL;
     }
+    if ( rddIteratorKey == null ) {
+      BlockDetails[] blocksInfo = GraphGenerator.blockInfo.get(meta.getBlocksID());
+      if (blocksInfo.length == 0) {
+        //if isDirectLoad = true, and partition number > file num
+        //then blocksInfo will get empty in some partition processing, so just return
+        setOutputDone();
+        return false;
+      }
 
-    BlockDetails[] blocksInfo = GraphGenerator.blockInfo.get(meta.getBlocksID());
-    if (blocksInfo.length == 0) {
-      //if isDirectLoad = true, and partition number > file num
-      //then blocksInfo will get empty in some partition processing, so just return
-      setOutputDone();
-      return false;
-    }
-
-    if (numberOfNodes > blocksInfo.length) {
-      numberOfNodes = blocksInfo.length;
-    }
+      if (numberOfNodes > blocksInfo.length) {
+        numberOfNodes = blocksInfo.length;
+      }
 
-    //new the empty lists
-    for (int pos = 0; pos < numberOfNodes; pos++) {
-      threadBlockList.add(new ArrayList<BlockDetails>());
-    }
+      //new the empty lists
+      for (int pos = 0; pos < numberOfNodes; pos++) {
+        threadBlockList.add(new ArrayList<BlockDetails>());
+      }
 
-    //block balance to every thread
-    for (int pos = 0; pos < blocksInfo.length; ) {
-      for (int threadNum = 0; threadNum < numberOfNodes; threadNum++) {
-        if (pos < blocksInfo.length) {
-          threadBlockList.get(threadNum).add(blocksInfo[pos++]);
+      //block balance to every thread
+      for (int pos = 0; pos < blocksInfo.length; ) {
+        for (int threadNum = 0; threadNum < numberOfNodes; threadNum++) {
+          if (pos < blocksInfo.length) {
+            threadBlockList.get(threadNum).add(blocksInfo[pos++]);
+          }
         }
       }
+      LOGGER.info("*****************Started all csv reading***********");
+      startProcess(numberOfNodes);
+      LOGGER.info("*****************Completed all csv reading***********");
+      CarbonTimeStatisticsFactory.getLoadStatisticsInstance().recordCsvInputStepTime(
+              meta.getPartitionID(), System.currentTimeMillis());
+    } else {
+      scanRddIterator();
     }
-    LOGGER.info("*****************Started all csv reading***********");
-    startProcess(numberOfNodes);
-    LOGGER.info("*****************Completed all csv reading***********");
-    CarbonTimeStatisticsFactory.getLoadStatisticsInstance().recordCsvInputStepTime(
-        meta.getPartitionID(), System.currentTimeMillis());
     setOutputDone();
     return false;
   }
 
+  private void scanRddIterator() throws RuntimeException {
+    Iterator<String[]> iterator = RddInputUtils.getAndRemove(rddIteratorKey);
+    if (iterator != null) {
+      try{
+        while(iterator.hasNext()){
+          putRow(data.outputRowMeta, iterator.next());
```

Comment:
One suggestion:
Instead of calling putRow here, perhaps we can execute more steps here until reaching merge step which need to wait. 
I suggest to raise another PR to implement this improvement.